tchannel/peer_heap.py

```diff
@@ -67,7 +84,28 @@ def pop_peer(self):
         return heap.pop(self)
 
     def push_peer(self, peer):
-        """Push a new peer into the heap"""
+        """Push a new peer into the heap
+
+        Order is equal to peer heap's current order plus random value within
+        the current peer size. It solves two problems when peers are in
+        same score:
+
+        1. 'dead peer' caused by random problem. If use fully random value, and
+        one peer is very unlock to get the worst value, as the peer selection
+        keeps working, that peer may not be selected anymore.
+
+        2. deterministic problem. If we just use push order, then all the
+        tchannel instance will follow the same pattern. If they all get the
+        same list of hyperbahn nodes, the peers that they pick to talk to
+        hyperbahn will be same. In that situation, it will cause the hyperbahn
+        node received very imbalance requests from tchannel clients.
+
+        All in all, it will keep certain level randomization but at the same
+        time make the peer score not deterministic among different tchannel
+        instance. For
+        """
+        self.order += 1
+        peer.order = self.order + random.randint(0, self.size())
```

Comment:
So, if you write a test:

    peers = [p1, p2, p3, p4]  # where p{1,2,3,4} have the same score
    for p in peers:
        heap.push(p)

    assert heap.peek_peer() == p1  # this will always be true

This test will always pass even though this is something you're trying to avoid.

It will always pass because even though orders for the four peers aren't 1, 2, 3, 4, `p1.order` is always smaller than order for p2, p3, or p4.